clang/include/clang/AST/Type.h

```diff
@@ -721,7 +721,9 @@ class Qualifiers {
            // to implicitly cast into the default address space.
            (A == LangAS::Default &&
             (B == LangAS::cuda_constant || B == LangAS::cuda_device ||
-             B == LangAS::cuda_shared));
+             B == LangAS::cuda_shared)) ||
+           // Otherwise, assume the default address space is compatible.
+           (A == LangAS::Default);
```

Comment:
Default is kind of broken, at least for OpenCL. Should avoid attaching any behavior to it

Reply:
Yeah I noted  there's actually a case where OpenCL touches this for function pointers. I might just have the OpenCL language call a different version of this. Overall the other targets just allow any AS to decay to default and that's likely the behavior we want when not bound by OpenCL semantics.

Reply:
@jhuber6, I agree with you, but you might want to check with @AnastasiaStulova.

We had a couple of lengthy discussions about re-using OpenCL attributes in SYCL mode ([here](https://reviews.llvm.org/D80932) and [here](https://github.com/intel/llvm/pull/1039/files)), but if I recall it correctly the conclusion was that OpenCL attributes inherit OpenCL semantics in non-OpenCL modes as well. The solution for SYCL mode was adding new attributes ([review](https://reviews.llvm.org/D89909)).

@Naghasan, FYI.

Reply:
> that's likely the behavior we want when not bound by OpenCL semantics.

I kind of agree, but that's up to the target to tell if this makes sense IMO. My mental model of what `Default` means in the case of CUDA/HIP/SYCL/OpenMP is "the address of a variable maps to the flat address space of the target".

My understanding here is you want a bypass for target address spaces to decay into default. While I think this is desirable in general, this disregards potential target limitations. The one I have in mind is for SPIR/SPIR-V, the target address space for constant shouldn't be allowed to decay into `Default`. There is some related discussions [here](https://discourse.llvm.org/t/rfc-improved-address-space-conversion-semantics-for-targets-and-language-dialects/51365) as well.